File: Leetcode/0208 Implement Trie (Prefix Tree).java
class Trie {
    private final Node root;

    public Trie() {
        root = new Node();
        root.isWordEnd = true;
    }

    public void insert(String word) {
        var currNode = root;
        for (int i = 0; i < word.length(); ++i) {
            int charIndex = word.charAt(i) - 'a';
            if (currNode.children[charIndex] == null) {
                currNode.children[charIndex] = new Node();
            }
            currNode = currNode.children[charIndex];
        }
        currNode.isWordEnd = true;
    }

    public boolean search(String word) {
        var currNode = root;
        for (int i = 0; currNode != null && i < word.length(); ++i) {
            int charIndex = word.charAt(i) - 'a';
            currNode = currNode.children[charIndex];
        }
        return currNode != null && currNode.isWordEnd;
    }

    public boolean startsWith(String prefix) {
        var currNode = root;
        for (int i = 0; currNode != null && i < prefix.length(); ++i) {
            int charIndex = prefix.charAt(i) - 'a';
            currNode = currNode.children[charIndex];
        }
        return currNode != null;
    }

    private static final int ALPHABET_SIZE = 26;

    private static class Node {
        boolean isWordEnd;
        Node[] children = new Node[ALPHABET_SIZE];
    }
}

File: Leetcode/0208 Implement Trie (Prefix Tree).ts
class TrieNode {
    children: Map<string, TrieNode>
    isEndOfWord: boolean

    constructor() {
        this.children = new Map();
        this.isEndOfWord = false;
    }
}

class Trie {
    root: TrieNode

    constructor() {
        this.root = new TrieNode();
    }

    insert(word: string): void {
        let curr = this.root;
        for (const char of word) {
            if (!curr.children.has(char)) {
                curr.children.set(char, new TrieNode());
            }
            curr = curr.children.get(char) as TrieNode;
        }
        curr.isEndOfWord = true;
    }

    search(word: string): boolean {
        let curr: TrieNode | undefined = this.root;
        for (let i = 0; i < word.length && curr != null; i++) {
            curr = curr.children.get(word[i]);
        }
        return curr != null && curr.isEndOfWord;
    }

    startsWith(prefix: string): boolean {
        let curr: TrieNode | undefined = this.root;
        for (let i = 0; i < prefix.length && curr != null; i++) {
            curr = curr.children.get(prefix[i]);
        }
        return curr != null;
    }
}
